File: Sources/SpeechX/AppState.swift
import Foundation
import Combine
import AppKit

private enum DefaultsKey: String {
    case selectedModel           = "selected_model"
    case selectedLanguage        = "selected_language"
    case selectedHotkey          = "selected_hotkey"
    case selectedLLMProvider     = "selected_llm_provider"
    case selectedGroqModel       = "selected_groq_model"
    case selectedOpenRouterModel = "selected_openrouter_model"
    case correctionModeEnabled   = "correction_mode_enabled"
    case grammarCorrectionEnabled = "grammar_correction_enabled"
    case codeMixEnabled          = "code_mix_enabled"
    case selectedCodeMix         = "selected_code_mix"
    case targetLanguageEnabled   = "target_language_enabled"
    case selectedTargetLanguage  = "selected_target_language"
    case feedbackSoundName       = "feedback_sound_name"
    case favoriteFocusWords      = "favorite_focus_words"
    case selectedAudioDeviceUID  = "selected_audio_device_uid"
    case customSystemPrompt      = "custom_system_prompt"
    case focusWords              = "focus_words"
    case totalWordsDictated      = "total_words_dictated"
    case currentStreakDays       = "current_streak_days"
    case averageWPM              = "average_wpm"
}

private extension UserDefaults {
    func string(forKey key: DefaultsKey) -> String? { string(forKey: key.rawValue) }
    func bool(forKey key: DefaultsKey) -> Bool     { bool(forKey: key.rawValue) }
    func set(_ value: Any?, forKey key: DefaultsKey) { set(value, forKey: key.rawValue) }
}

enum RecordingState {
    case idle
    case recording
    case transcribing
    case error(String)
}

struct TranscriptEntry: Identifiable, Equatable {
    let id = UUID()
    let timestamp: Date
    let raw: String
    let processed: String?

    /// What was actually typed into the focused app.
    var typed: String { processed ?? raw }

    /// True iff an LLM step ran and changed the text. Drives the "Copy raw"
    /// alternate menu item — there's nothing to compare against otherwise.
    var hasLLMProcessing: Bool {
        guard let processed else { return false }
        return processed != raw
    }
}

enum HotkeyOption: String, CaseIterable, Identifiable {
    case rightOption  = "right_option"
    case leftOption   = "left_option"
    case rightCommand = "right_command"
    case leftCommand  = "left_command"
    case fn           = "fn"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .rightOption:  return "Right Option (⌥)"
        case .leftOption:   return "Left Option (⌥)"
        case .rightCommand: return "Right Command (⌘)"
        case .leftCommand:  return "Left Command (⌘)"
        case .fn:           return "Fn (🌐)"
        }
    }

    var keyCode: UInt16 {
        switch self {
        case .rightOption:  return 61
        case .leftOption:   return 58
        case .rightCommand: return 54
        case .leftCommand:  return 55
        case .fn:           return 63
        }
    }

    var modifierFlag: NSEvent.ModifierFlags {
        switch self {
        case .rightOption, .leftOption:   return .option
        case .rightCommand, .leftCommand: return .command
        case .fn:                         return .function
        }
    }
}

class AppState: ObservableObject {
    @Published var recordingState: RecordingState = .idle
    @Published var deepgramAPIKey: String = ""
    @Published var lastTranscript: String = ""
    /// Running transcript shown in the recording overlay while capture is active.
    /// Updated from Deepgram interim + final results; cleared on each new session.
    @Published var liveTranscript: String = ""
    @Published var availableModels: [DeepgramModel] = []

    /// Newest first. Capped at `transcriptHistoryLimit`.
    @Published var transcriptHistory: [TranscriptEntry] = []
    private let transcriptHistoryLimit = 20

    func recordTranscript(raw: String, processed: String?) {
        let entry = TranscriptEntry(timestamp: Date(), raw: raw, processed: processed)
        
        // Update stats
        let words = entry.typed.split(separator: " ").count
        totalWordsDictated += words
        
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.transcriptHistory.insert(entry, at: 0)
            if self.transcriptHistory.count > self.transcriptHistoryLimit {
                self.transcriptHistory.removeLast(self.transcriptHistory.count - self.transcriptHistoryLimit)
            }
        }
    }

    func clearTranscriptHistory() {
        DispatchQueue.main.async { [weak self] in
            self?.transcriptHistory.removeAll()
        }
    }

    /// Short user-facing message that auto-clears after a few seconds. Drives the
    /// menu-bar warning icon flash and tooltip.
    @Published var transientError: String? = nil
    private var transientErrorClearWorkItem: DispatchWorkItem?

    /// Surface a non-fatal error to the user. Logs it and triggers a 4s flash.
    func reportError(_ message: String) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.transientError = message
            self.transientErrorClearWorkItem?.cancel()
            let work = DispatchWorkItem { [weak self] in
                self?.transientError = nil
            }
            self.transientErrorClearWorkItem = work
            DispatchQueue.main.asyncAfter(deadline: .now() + 4, execute: work)
        }
    }

    @Published var selectedModel: String {
        didSet { UserDefaults.standard.set(selectedModel, forKey: .selectedModel) }
    }

    @Published var selectedLanguage: String {
        didSet { UserDefaults.standard.set(selectedLanguage, forKey: .selectedLanguage) }
    }

    @Published var selectedHotkey: HotkeyOption {
        didSet { UserDefaults.standard.set(selectedHotkey.rawValue, forKey: .selectedHotkey) }
    }

    @Published var selectedLLMProvider: LLMProvider {
        didSet { UserDefaults.standard.set(selectedLLMProvider.rawValue, forKey: .selectedLLMProvider) }
    }

    @Published var groqAPIKey: String = ""
    @Published var availableGroqModels: [LLMModel] = []

    @Published var selectedGroqModel: String {
        didSet { UserDefaults.standard.set(selectedGroqModel, forKey: .selectedGroqModel) }
    }

    @Published var openRouterAPIKey: String = ""
    @Published var availableOpenRouterModels: [LLMModel] = []

    @Published var selectedOpenRouterModel: String {
        didSet { UserDefaults.standard.set(selectedOpenRouterModel, forKey: .selectedOpenRouterModel) }
    }

    @Published var correctionModeEnabled: Bool {
        didSet { UserDefaults.standard.set(correctionModeEnabled, forKey: .correctionModeEnabled) }
    }

    @Published var grammarCorrectionEnabled: Bool {
        didSet { UserDefaults.standard.set(grammarCorrectionEnabled, forKey: .grammarCorrectionEnabled) }
    }

    @Published var codeMixEnabled: Bool {
        didSet { UserDefaults.standard.set(codeMixEnabled, forKey: .codeMixEnabled) }
    }

    @Published var selectedCodeMix: String {
        didSet { UserDefaults.standard.set(selectedCodeMix, forKey: .selectedCodeMix) }
    }

    @Published var targetLanguageEnabled: Bool {
        didSet { UserDefaults.standard.set(targetLanguageEnabled, forKey: .targetLanguageEnabled) }
    }

    @Published var selectedTargetLanguage: String {
        didSet { UserDefaults.standard.set(selectedTargetLanguage, forKey: .selectedTargetLanguage) }
    }

    /// System sound played on record start/stop. Empty string = muted.
    @Published var feedbackSoundName: String {
        didSet { UserDefaults.standard.set(feedbackSoundName, forKey: .feedbackSoundName) }
    }

    /// Free-form text prepended to the LLM system prompt — lets users bias output
    /// (tone, terminology, glossary). Empty = disabled.
    @Published var customSystemPrompt: String {
        didSet { UserDefaults.standard.set(customSystemPrompt, forKey: .customSystemPrompt) }
    }

    /// Focus-words dictionary — newline/comma-separated terms (names, emails, jargon) whose exact
    /// spelling matters. Fed to Deepgram as keyterms (Nova-3 only) and to the LLM as a glossary.
    /// Empty = disabled.
    @Published var focusWords: String {
        didSet { UserDefaults.standard.set(focusWords, forKey: .focusWords) }
    }

    @Published var totalWordsDictated: Int = 0 {
        didSet { UserDefaults.standard.set(totalWordsDictated, forKey: .totalWordsDictated) }
    }
    
    @Published var currentStreakDays: Int = 0 {
        didSet { UserDefaults.standard.set(currentStreakDays, forKey: .currentStreakDays) }
    }
    
    @Published var averageWPM: Int = 0 {
        didSet { UserDefaults.standard.set(averageWPM, forKey: .averageWPM) }
    }

    /// Parsed, de-duplicated focus-word trigger keys. Drives the Deepgram keyterm query.
    var focusWordTerms: [String] { FocusWordsDictionary.keys(focusWords) }

    /// Lowercased trigger keys the user starred in the Dictionary tab. Purely
    /// organizational (pinned to the top of the list) — no recognition effect.
    @Published var favoriteFocusWords: Set<String> {
        didSet { UserDefaults.standard.set(Array(favoriteFocusWords), forKey: .favoriteFocusWords) }
    }

    /// AVCaptureDevice.uniqueID of the chosen mic. Empty string = system default.
    @Published var selectedAudioDeviceUID: String {
        didSet { UserDefaults.standard.set(selectedAudioDeviceUID, forKey: .selectedAudioDeviceUID) }
    }

    /// Populated from `refreshAudioDevices()`. Not persisted — devices come and go.
    @Published var availableAudioDevices: [AudioInputDevice] = []

    func refreshAudioDevices() {
        let devices = AudioEngine.availableInputDevices()
        DispatchQueue.main.async { [weak self] in
            self?.availableAudioDevices = devices
        }
    }

    let audioEngine: AudioEngine = AudioEngine()
    let deepgramService: DeepgramService
    let llmService: LLMService = LLMService()
    let textInjector: TextInjector = TextInjector()
    /// Watches for the user typing over a word SpeechX just injected, and learns
    /// the corrected spelling into the focus-words dictionary.
    let typeOverWatcher: TypeOverWatcher = TypeOverWatcher()
    let keychainService: KeychainService = KeychainService()
    let audioMuter: SystemAudioMuter = SystemAudioMuter()

    init() {
        self.deepgramService = DeepgramService()
        self.deepgramAPIKey = keychainService.retrieve(key: "deepgram_api_key") ?? ""

        let storedModel = UserDefaults.standard.string(forKey: .selectedModel) ?? "nova-3-general"
        self.selectedModel = storedModel

        let storedLanguage = UserDefaults.standard.string(forKey: .selectedLanguage) ?? "en-US"
        self.selectedLanguage = storedLanguage

        let storedHotkey = UserDefaults.standard.string(forKey: .selectedHotkey) ?? ""
        self.selectedHotkey = HotkeyOption(rawValue: storedHotkey) ?? .rightOption

        let storedProvider = UserDefaults.standard.string(forKey: .selectedLLMProvider) ?? ""
        self.selectedLLMProvider = LLMProvider(rawValue: storedProvider) ?? .groq

        self.groqAPIKey = keychainService.retrieve(key: LLMProvider.groq.keychainKey) ?? ""
        self.selectedGroqModel = UserDefaults.standard.string(forKey: .selectedGroqModel) ?? ""

        self.openRouterAPIKey = keychainService.retrieve(key: LLMProvider.openRouter.keychainKey) ?? ""
        self.selectedOpenRouterModel = UserDefaults.standard.string(forKey: .selectedOpenRouterModel) ?? ""

        self.correctionModeEnabled = UserDefaults.standard.bool(forKey: .correctionModeEnabled)
        self.grammarCorrectionEnabled = UserDefaults.standard.bool(forKey: .grammarCorrectionEnabled)
        self.codeMixEnabled = UserDefaults.standard.bool(forKey: .codeMixEnabled)
        self.selectedCodeMix = UserDefaults.standard.string(forKey: .selectedCodeMix) ?? ""
        self.targetLanguageEnabled = UserDefaults.standard.bool(forKey: .targetLanguageEnabled)

        // Migration: code-mix styles used to live in the target-language picker.
        // They now belong only to the Code-Mix Input toggle.
        let storedTarget = UserDefaults.standard.string(forKey: .selectedTargetLanguage) ?? "English"
        let codeMixStyles: Set<String> = [
            "Hinglish", "Tanglish", "Benglish", "Kanglish", "Tenglish",
            "Minglish", "Punglish", "Spanglish", "Franglais", "Portuñol",
            "Chinglish", "Japlish", "Konglish", "Arabizi", "Sheng", "Camfranglais"
        ]
        self.selectedTargetLanguage = codeMixStyles.contains(storedTarget) ? "English" : storedTarget

        self.feedbackSoundName = UserDefaults.standard.string(forKey: .feedbackSoundName) ?? "Tink"
        self.customSystemPrompt = UserDefaults.standard.string(forKey: .customSystemPrompt) ?? ""
        self.focusWords = UserDefaults.standard.string(forKey: .focusWords) ?? ""
        self.favoriteFocusWords = Set(
            UserDefaults.standard.stringArray(forKey: DefaultsKey.favoriteFocusWords.rawValue) ?? []
        )
        
        self.totalWordsDictated = UserDefaults.standard.integer(forKey: DefaultsKey.totalWordsDictated.rawValue)
        self.currentStreakDays = UserDefaults.standard.integer(forKey: DefaultsKey.currentStreakDays.rawValue)
        self.averageWPM = UserDefaults.standard.integer(forKey: DefaultsKey.averageWPM.rawValue)

        self.selectedAudioDeviceUID = UserDefaults.standard.string(forKey: .selectedAudioDeviceUID) ?? ""

        self.deepgramService.onPartialTranscript = { [weak self] text in
            // Service already hops to the main queue.
            self?.liveTranscript = text
        }

        self.typeOverWatcher.onCorrection = { [weak self] corrected, original in
            self?.learnWord(corrected, original: original)
        }
    }

    // MARK: - Keyword learning (type-over)

    /// Start watching the focused field for a correction of the text just injected.
    func noteInjection(_ text: String) {
        typeOverWatcher.watch(injected: text)
    }

    /// Auto-add a learned spelling to the focus-words dictionary (a spelling lock
    /// that biases Deepgram next time). `original` is the word that was typed and
    /// corrected away from: any existing bare focus word that's `original` or a close
    /// spelling-variant of it is removed first, so successive re-spellings of the same
    /// word collapse to just the latest rather than accumulating.
    ///
    /// Only single bare-word lines are ever removed — "trigger : replacement" entries
    /// and multi-word phrases are left untouched.
    func learnWord(_ word: String, original: String) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            let term = word.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !term.isEmpty else { return }
            let originalLC = original.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()

            var lines = self.focusWords.isEmpty ? [] : self.focusWords.components(separatedBy: "\n")
            if !originalLC.isEmpty {
                lines.removeAll { line in
                    let w = line.trimmingCharacters(in: .whitespacesAndNewlines)
                    guard !w.isEmpty,
                          !w.contains(":"), !w.contains(" "), !w.contains("\t"), // bare single word only
                          w.caseInsensitiveCompare(term) != .orderedSame          // never the word we're adding
                    else { return false }
                    let wLC = w.lowercased()
                    let dist = TypeOverDetector.levenshtein(wLC, originalLC)
                    return dist <= max(2, max(wLC.count, originalLC.count) / 3)
                }
            }
            let exists = lines.contains {
                $0.trimmingCharacters(in: .whitespacesAndNewlines).caseInsensitiveCompare(term) == .orderedSame
            }
            if !exists { lines.append(term) }
            self.focusWords = lines.joined(separator: "\n")
        }
    }

    // MARK: - Helpers for the selected LLM provider

    var currentLLMAPIKey: String {
        switch selectedLLMProvider {
        case .groq:       return groqAPIKey
        case .openRouter: return openRouterAPIKey
        }
    }

    var currentLLMModel: String {
        switch selectedLLMProvider {
        case .groq:       return selectedGroqModel
        case .openRouter: return selectedOpenRouterModel
        }
    }
}

File: windows/SpeechX/Core/AppState.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Windows;
using SpeechX.Services;

namespace SpeechX.Core;

/// <summary>
/// Central observable application state and the home of the shared services. Port of the macOS
/// AppState (ObservableObject -> INotifyPropertyChanged). Persisted preferences go through
/// SettingsStore; secret API keys go through CredentialStore (DPAPI).
/// </summary>
public sealed class AppState : INotifyPropertyChanged
{
    // Persisted-preference keys (match the macOS DefaultsKey raw values).
    private static class Keys
    {
        public const string SelectedModel = "selected_model";
        public const string SelectedLanguage = "selected_language";
        public const string SelectedHotkey = "selected_hotkey";
        public const string SelectedLlmProvider = "selected_llm_provider";
        public const string SelectedGroqModel = "selected_groq_model";
        public const string SelectedOpenRouterModel = "selected_openrouter_model";
        public const string CorrectionModeEnabled = "correction_mode_enabled";
        public const string GrammarCorrectionEnabled = "grammar_correction_enabled";
        public const string CodeMixEnabled = "code_mix_enabled";
        public const string SelectedCodeMix = "selected_code_mix";
        public const string TargetLanguageEnabled = "target_language_enabled";
        public const string SelectedTargetLanguage = "selected_target_language";
        public const string FeedbackSoundName = "feedback_sound_name";
        public const string SelectedAudioDeviceUid = "selected_audio_device_uid";
        public const string CustomSystemPrompt = "custom_system_prompt";
        public const string FocusWords = "focus_words";
        public const string FavoriteFocusWords = "favorite_focus_words";
        public const string AutoUpdateEnabled = "auto_update_enabled";
    }

    // Shared services.
    public AudioEngine AudioEngine { get; } = new();
    public DeepgramService DeepgramService { get; } = new();
    public LlmService LlmService { get; } = new();
    public TextInjector TextInjector { get; } = new();
    /// <summary>Watches for the user typing over a word SpeechX just injected, and learns the
    /// corrected spelling into the focus-words dictionary.</summary>
    public TypeOverWatcher TypeOverWatcher { get; } = new();
    public SystemAudioMuter AudioMuter { get; } = new();
    public SettingsStore Settings { get; } = new();
    public CredentialStore Credentials { get; } = new();

    private const int TranscriptHistoryLimit = 20;
    private int _errorGeneration;

    public AppState()
    {
        DeepgramApiKey = Credentials.Retrieve("deepgram_api_key") ?? "";

        _selectedModel = Settings.GetString(Keys.SelectedModel) ?? "nova-3-general";
        _selectedLanguage = Settings.GetString(Keys.SelectedLanguage) ?? "en-US";
        _selectedHotkey = HotkeyOptionExtensions.FromRawValue(Settings.GetString(Keys.SelectedHotkey));
        _selectedLlmProvider = LlmProviderExtensions.FromRawValue(Settings.GetString(Keys.SelectedLlmProvider));

        GroqApiKey = Credentials.Retrieve(LlmProvider.Groq.CredentialKey()) ?? "";
        _selectedGroqModel = Settings.GetString(Keys.SelectedGroqModel) ?? "";

        OpenRouterApiKey = Credentials.Retrieve(LlmProvider.OpenRouter.CredentialKey()) ?? "";
        _selectedOpenRouterModel = Settings.GetString(Keys.SelectedOpenRouterModel) ?? "";

        _correctionModeEnabled = Settings.GetBool(Keys.CorrectionModeEnabled);
        _grammarCorrectionEnabled = Settings.GetBool(Keys.GrammarCorrectionEnabled);
        _codeMixEnabled = Settings.GetBool(Keys.CodeMixEnabled);
        _selectedCodeMix = Settings.GetString(Keys.SelectedCodeMix) ?? "";
        _targetLanguageEnabled = Settings.GetBool(Keys.TargetLanguageEnabled);

        // Migration: code-mix styles used to live in the target-language picker.
        var storedTarget = Settings.GetString(Keys.SelectedTargetLanguage) ?? "English";
        var codeMixStyles = new HashSet<string>
        {
            "Hinglish", "Tanglish", "Benglish", "Kanglish", "Tenglish",
            "Minglish", "Punglish", "Spanglish", "Franglais", "Portuñol",
            "Chinglish", "Japlish", "Konglish", "Arabizi", "Sheng", "Camfranglais",
        };
        _selectedTargetLanguage = codeMixStyles.Contains(storedTarget) ? "English" : storedTarget;

        _feedbackSoundName = Settings.GetString(Keys.FeedbackSoundName) ?? "Asterisk";
        _customSystemPrompt = Settings.GetString(Keys.CustomSystemPrompt) ?? "";
        _focusWords = Settings.GetString(Keys.FocusWords) ?? "";
        _favoriteFocusWords = new HashSet<string>(
            (Settings.GetString(Keys.FavoriteFocusWords) ?? "")
                .Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries));
        _selectedAudioDeviceUid = Settings.GetString(Keys.SelectedAudioDeviceUid) ?? "";

        // Auto-update defaults ON; only an explicit "false" disables it.
        _autoUpdateEnabled = Settings.GetString(Keys.AutoUpdateEnabled) != "false";

        DeepgramService.OnPartialTranscript = text => OnUi(() => LiveTranscript = text);

        TypeOverWatcher.OnCorrection = (corrected, original) => LearnWord(corrected, original);
    }

    // MARK: - Keyword learning (type-over)

    /// <summary>Start watching the focused field for a correction of the text just injected.</summary>
    public void NoteInjection(string text) => TypeOverWatcher.Watch(text);

    /// <summary>
    /// Auto-add a learned spelling to the focus-words dictionary (a spelling lock that biases Deepgram
    /// next time). <paramref name="original"/> is the word that was typed and corrected away from: any
    /// existing bare focus word that's <paramref name="original"/> or a close spelling-variant of it is
    /// removed first, so successive re-spellings of the same word collapse to just the latest rather
    /// than accumulating. Only single bare-word lines are ever removed — "trigger : replacement"
    /// entries and multi-word phrases are left untouched.
    /// </summary>
    public void LearnWord(string word, string original)
    {
        OnUi(() =>
        {
            var term = word.Trim();
            if (term.Length == 0) return;
            var originalLC = original.Trim().ToLowerInvariant();

            var lines = string.IsNullOrEmpty(_focusWords)
                ? new List<string>()
                : _focusWords.Split('\n').ToList();

            if (originalLC.Length > 0)
            {
                lines.RemoveAll(line =>
                {
                    var w = line.Trim();
                    if (w.Length == 0) return false;
                    if (w.Contains(':') || w.Contains(' ') || w.Contains('\t')) return false; // bare single word only
                    if (string.Equals(w, term, StringComparison.OrdinalIgnoreCase)) return false; // never the word we're adding
                    var wLC = w.ToLowerInvariant();
                    var dist = TypeOverDetector.Levenshtein(wLC, originalLC);
                    return dist <= Math.Max(2, Math.Max(wLC.Length, originalLC.Length) / 3);
                });
            }

            var exists = lines.Any(l => string.Equals(l.Trim(), term, StringComparison.OrdinalIgnoreCase));
            if (!exists) lines.Add(term);
            FocusWords = string.Join("\n", lines);
        });
    }

    // MARK: - Volatile UI state

    private RecordingState _recordingState = RecordingState.Idle;
    public RecordingState RecordingState { get => _recordingState; set => SetField(ref _recordingState, value); }

    public string DeepgramApiKey { get; set; } = "";

    public string LastTranscript { get; set; } = "";

    private string _liveTranscript = "";
    public string LiveTranscript { get => _liveTranscript; set => SetField(ref _liveTranscript, value); }

    private IReadOnlyList<DeepgramModel> _availableModels = Array.Empty<DeepgramModel>();
    public IReadOnlyList<DeepgramModel> AvailableModels { get => _availableModels; set => SetField(ref _availableModels, value); }

    private List<TranscriptEntry> _transcriptHistory = new();
    public IReadOnlyList<TranscriptEntry> TranscriptHistory => _transcriptHistory;
        public int TotalWordsDictated
        {
            get => int.TryParse(Settings.GetString("TotalWordsDictated"), out var v) ? v : 0;
            set { Settings.SetString("TotalWordsDictated", value.ToString()); OnPropertyChanged(); }
        }
        
        public int CurrentStreakDays
        {
            get => int.TryParse(Settings.GetString("CurrentStreakDays"), out var v) ? v : 0;
            set { Settings.SetString("CurrentStreakDays", value.ToString()); OnPropertyChanged(); }
        }
        
        public int AverageWPM
        {
            get => int.TryParse(Settings.GetString("AverageWPM"), out var v) ? v : 0;
            set { Settings.SetString("AverageWPM", value.ToString()); OnPropertyChanged(); }
        }

    public void RecordTranscript(string raw, string? processed)
    {
        OnUi(() =>
        {
            _transcriptHistory.Insert(0, new TranscriptEntry(DateTime.Now, raw, processed));
            if (_transcriptHistory.Count > TranscriptHistoryLimit)
                _transcriptHistory.RemoveRange(TranscriptHistoryLimit, _transcriptHistory.Count - TranscriptHistoryLimit);
            OnPropertyChanged(nameof(TranscriptHistory));
        });
    }

    public void ClearTranscriptHistory()
    {
        OnUi(() =>
        {
            _transcriptHistory.Clear();
            OnPropertyChanged(nameof(TranscriptHistory));
        });
    }

    private string? _transientError;
    public string? TransientError { get => _transientError; private set => SetField(ref _transientError, value); }

    /// <summary>Surface a non-fatal error to the user; auto-clears after 4 seconds.</summary>
    public void ReportError(string message)
    {
        OnUi(() =>
        {
            TransientError = message;
            int gen = ++_errorGeneration;
            _ = Task.Delay(4000).ContinueWith(_ => OnUi(() =>
            {
                if (_errorGeneration == gen) TransientError = null;
            }), TaskScheduler.Default);
        });
    }

    // MARK: - Persisted preferences

    private string _selectedModel;
    public string SelectedModel
    {
        get => _selectedModel;
        set { if (SetField(ref _selectedModel, value)) Settings.SetString(Keys.SelectedModel, value); }
    }

    private string _selectedLanguage;
    public string SelectedLanguage
    {
        get => _selectedLanguage;
        set { if (SetField(ref _selectedLanguage, value)) Settings.SetString(Keys.SelectedLanguage, value); }
    }

    private HotkeyOption _selectedHotkey;
    public HotkeyOption SelectedHotkey
    {
        get => _selectedHotkey;
        set { if (SetField(ref _selectedHotkey, value)) Settings.SetString(Keys.SelectedHotkey, value.ToRawValue()); }
    }

    private LlmProvider _selectedLlmProvider;
    public LlmProvider SelectedLlmProvider
    {
        get => _selectedLlmProvider;
        set { if (SetField(ref _selectedLlmProvider, value)) Settings.SetString(Keys.SelectedLlmProvider, value.ToRawValue()); }
    }

    public string GroqApiKey { get; set; } = "";
    private IReadOnlyList<LlmModel> _availableGroqModels = Array.Empty<LlmModel>();
    public IReadOnlyList<LlmModel> AvailableGroqModels { get => _availableGroqModels; set => SetField(ref _availableGroqModels, value); }

    private string _selectedGroqModel;
    public string SelectedGroqModel
    {
        get => _selectedGroqModel;
        set { if (SetField(ref _selectedGroqModel, value)) Settings.SetString(Keys.SelectedGroqModel, value); }
    }

    public string OpenRouterApiKey { get; set; } = "";
    private IReadOnlyList<LlmModel> _availableOpenRouterModels = Array.Empty<LlmModel>();
    public IReadOnlyList<LlmModel> AvailableOpenRouterModels { get => _availableOpenRouterModels; set => SetField(ref _availableOpenRouterModels, value); }

    private string _selectedOpenRouterModel;
    public string SelectedOpenRouterModel
    {
        get => _selectedOpenRouterModel;
        set { if (SetField(ref _selectedOpenRouterModel, value)) Settings.SetString(Keys.SelectedOpenRouterModel, value); }
    }

    private bool _correctionModeEnabled;
    public bool CorrectionModeEnabled
    {
        get => _correctionModeEnabled;
        set { if (SetField(ref _correctionModeEnabled, value)) Settings.SetBool(Keys.CorrectionModeEnabled, value); }
    }

    private bool _grammarCorrectionEnabled;
    public bool GrammarCorrectionEnabled
    {
        get => _grammarCorrectionEnabled;
        set { if (SetField(ref _grammarCorrectionEnabled, value)) Settings.SetBool(Keys.GrammarCorrectionEnabled, value); }
    }

    private bool _codeMixEnabled;
    public bool CodeMixEnabled
    {
        get => _codeMixEnabled;
        set { if (SetField(ref _codeMixEnabled, value)) Settings.SetBool(Keys.CodeMixEnabled, value); }
    }

    private string _selectedCodeMix;
    public string SelectedCodeMix
    {
        get => _selectedCodeMix;
        set { if (SetField(ref _selectedCodeMix, value)) Settings.SetString(Keys.SelectedCodeMix, value); }
    }

    private bool _targetLanguageEnabled;
    public bool TargetLanguageEnabled
    {
        get => _targetLanguageEnabled;
        set { if (SetField(ref _targetLanguageEnabled, value)) Settings.SetBool(Keys.TargetLanguageEnabled, value); }
    }

    private string _selectedTargetLanguage;
    public string SelectedTargetLanguage
    {
        get => _selectedTargetLanguage;
        set { if (SetField(ref _selectedTargetLanguage, value)) Settings.SetString(Keys.SelectedTargetLanguage, value); }
    }

    private string _feedbackSoundName;
    public string FeedbackSoundName
    {
        get => _feedbackSoundName;
        set { if (SetField(ref _feedbackSoundName, value)) Settings.SetString(Keys.FeedbackSoundName, value); }
    }

    private string _customSystemPrompt;
    public string CustomSystemPrompt
    {
        get => _customSystemPrompt;
        set { if (SetField(ref _customSystemPrompt, value)) Settings.SetString(Keys.CustomSystemPrompt, value); }
    }

    private string _focusWords;
    /// <summary>
    /// Focus-words dictionary — newline/comma-separated terms (names, emails, jargon) whose exact
    /// spelling matters. Fed to Deepgram as keyterms (Nova-3 only) and to the LLM as a glossary.
    /// Empty = disabled.
    /// </summary>
    public string FocusWords
    {
        get => _focusWords;
        set { if (SetField(ref _focusWords, value)) Settings.SetString(Keys.FocusWords, value); }
    }

    /// <summary>Parsed, de-duplicated focus-word trigger keys. Drives the Deepgram keyterm query.</summary>
    public IReadOnlyList<string> FocusWordTerms => FocusWordsDictionary.Keys(_focusWords);

    private HashSet<string> _favoriteFocusWords;
    /// <summary>
    /// Lowercased dictionary keys the user starred in the Dictionary tab. Purely a UI pin (favorites
    /// sort to the top of the entry list) — has no effect on recognition. Mirrors the macOS
    /// favorite_focus_words set.
    /// </summary>
    public IReadOnlySet<string> FavoriteFocusWords => _favoriteFocusWords;

    public bool IsFavoriteFocusWord(string key) => _favoriteFocusWords.Contains(key.ToLowerInvariant());

    public void SetFavoriteFocusWord(string key, bool favorite)
    {
        var k = key.ToLowerInvariant();
        bool changed = favorite ? _favoriteFocusWords.Add(k) : _favoriteFocusWords.Remove(k);
        if (!changed) return;
        Settings.SetString(Keys.FavoriteFocusWords, string.Join("\n", _favoriteFocusWords));
        OnPropertyChanged(nameof(FavoriteFocusWords));
    }

    private string _selectedAudioDeviceUid;
    public string SelectedAudioDeviceUid
    {
        get => _selectedAudioDeviceUid;
        set { if (SetField(ref _selectedAudioDeviceUid, value)) Settings.SetString(Keys.SelectedAudioDeviceUid, value); }
    }

    private bool _autoUpdateEnabled;
    public bool AutoUpdateEnabled
    {
        get => _autoUpdateEnabled;
        set { if (SetField(ref _autoUpdateEnabled, value)) Settings.SetBool(Keys.AutoUpdateEnabled, value); }
    }

    private IReadOnlyList<AudioInputDevice> _availableAudioDevices = Array.Empty<AudioInputDevice>();
    public IReadOnlyList<AudioInputDevice> AvailableAudioDevices { get => _availableAudioDevices; set => SetField(ref _availableAudioDevices, value); }

    public void RefreshAudioDevices()
    {
        var devices = AudioEngine.AvailableInputDevices();
        OnUi(() => AvailableAudioDevices = devices);
    }

    // MARK: - Helpers for the selected LLM provider

    public string CurrentLlmApiKey => SelectedLlmProvider switch
    {
        LlmProvider.Groq => GroqApiKey,
        LlmProvider.OpenRouter => OpenRouterApiKey,
        _ => GroqApiKey,
    };

    public string CurrentLlmModel => SelectedLlmProvider switch
    {
        LlmProvider.Groq => SelectedGroqModel,
        LlmProvider.OpenRouter => SelectedOpenRouterModel,
        _ => SelectedGroqModel,
    };

    // MARK: - INotifyPropertyChanged

    public event PropertyChangedEventHandler? PropertyChanged;

    private void OnPropertyChanged([CallerMemberName] string? name = null)
        => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));

    private bool SetField<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return false;
        field = value;
        OnPropertyChanged(name);
        return true;
    }

    /// <summary>Run on the WPF UI thread (or inline if already there / no app).</summary>
    internal static void OnUi(Action action)
    {
        var disp = Application.Current?.Dispatcher;
        if (disp == null || disp.CheckAccess()) action();
        else disp.BeginInvoke(action);
    }
}


